README.md
sandbox.lua
===========

A pure-lua solution for running untrusted Lua code.

The default behavior is restricting access to "dangerous" functions in Lua, such as `os.execute`.

It's possible to provide extra functions via the `options.env` parameter.

Infinite loops are prevented via the `debug` library.

Supported Lua versions:
======================

All the features of sandbox.lua work in the following Lua environments:


* PUC-Rio Lua 5.1 **allows execution of bytecode**, which is a huge limitation (see the bytecode section below)
* PUC-Rio Lua 5.2, 5.3, 5.4 have total support.
* LuaJIT is not protected against infinite loops (see the notes in `options.quota` below)

Usage
=====

Require the module like this:

``` lua
local sandbox = require 'sandbox'
```

Then you can use `sandbox.run` and `sandbox.protect`

### sandbox.run(code, options, ...)

`sandbox.run(code, options, ...)` sandboxes and executes `code` with the given `options` and extra params.

`code` must be a string with Lua code inside.

`options` is described below.

Any extra parameters will just be passed to the sandboxed function when executed, and available on the top-level scope via the `...` varargs parameters.

In other words, `sandbox.run(c, o, ...)` is equivalent to `sandbox.protect(c, o)(...)`.

Notice that if `code` throws an error, it is *NOT* captured by `sandbox.run`. Use `pcall` if you want your app to be immune to errors, like this:

``` lua
local ok, result = pcall(sandbox.run, 'error("this just throws an error")')
```

### sandbox.protect(code, options)

`sandbox.protect("lua code")` (or `sandbox("lua code")`) produces a sandboxed function, without executing it.

The resulting sandboxed function works as regular functions as long as they don't access any insecure features:

```lua
local sandboxed_f = sandbox(function() return 'hey' end)
local msg = sandboxed_f() -- msg is now 'hey'
```

Sandboxed options can not access unsafe Lua modules. (See the [source code](https://github.com/kikito/sandbox.lua/blob/master/sandbox.lua#L35) for a list)

When a sandboxed function tries to access an unsafe module, an error is produced.

```lua
local sf = sandbox.protect([[
  os.execute('rm -rf /') -- this will throw an error, no damage done
end
]])

sf() -- error: os.execute not found
```

Sandboxed code will eventually throw an error if it contains infinite loops (note: this feature is not available in LuaJIT):

```lua
local sf = sandbox.protect([[
  while true do end
]])

sf() -- error: quota exceeded
```

### Bytecode

It is possible to exit a sandbox using specially-crafted Lua bytecode. References:

* http://apocrypha.numin.it/talks/lua_bytecode_exploitation.pdf
* https://github.com/erezto/lua-sandbox-escape
* https://gist.github.com/corsix/6575486

Because of this, the sandbox deactivates bytecode in all the versions of Lua where it is possible:

* PUC-Rio Lua 5.2, 5.3, 5.4
* LuaJIT

In other words, _all except PUC-Rio Lua 5.1_.

** The sandbox can be exploited in PUC-Rio Lua 5.1 via bytecode **

The only reason we keep Lua 5.1 in the list of supported versions of Lua is because
sandboxing can help against users attempting to delete a file by mistake. _It does not provide
protection against malicious users_.

As a result we _strongly recommend updating to a more recent version when possible_.

### options.quota

Note: This feature is not available in LuaJIT

`sandbox.lua` prevents infinite loops from halting the program by hooking the `debug` library to the sandboxed function, and "counting instructions". When
the instructions reach a certain limit, an error is produced.

This limit can be tweaked via the `quota` option. But default, it is 500000.

It is not possible to exhaust the machine with infinite loops; the following will throw an error after invoking 500000 instructions:

``` lua
sandbox.run('while true do end') -- raise errors after 500000 instructions
sandbox.run('while true do end', {quota=10000}) -- raise error after 10000 instructions
```

If the quota is low enough, sandboxed code with too many calculations might fail:

``` lua
local code = [[
  local count = 1
  for i=1, 400 do count = count + 1 end
  return count
]]

sandbox.run(code, {quota=100}) -- raises error before the code ends
```

If you want to turn off the quota completely, pass `quota=false` instead.


### options.env

Use the `env` option to inject additional variables to the environment in which the sandboxed code is executed.

    local msg = sandbox.run('return foo', {env = {foo = 'This is a global var on the the environment'}})

The `env` variable will be used as an "index" by the sandbox environment, but it will *not* be modified at all (changes
to the environment are thus lost). The only way to "get information out" from the sandboxed environments are:

Through side effects, like writing to a database. You will have to provide the side-effects functions in `env`:

    local val = 1
    local env = { write_db = function(new_val) val = new_val end }
    sandbox.run('write_db(2)')
    assert(val = 2)

Through returned values:

    local env = { amount = 1 }
    local result = sandbox.run('return amount + 1', { env = env })
    assert(result = 2)


Installation
============

Just copy sandbox.lua wherever you need it.

Alternatively, you can use luarocks:

    luarocks install kikito/sandbox

License
=======

This library is released under the MIT license. See MIT-LICENSE.txt for details

Specs
=====

This project uses [busted](https://github.com/Olivine-Labs/busted) for its specs. In order to run them, install it and then:

```
cd /path/to/where/the/spec/folder/is
busted spec/*
```


CHANGELOG.md
# v1.0.1 (2021-01)

- Fix a bug in which the base environment wasn't overrideable with `false`

# v1.0.0 (2021-01)

- Added support for all major versions of PUC Rio Lua and LuaJIT
- Only Lua strings are admitted now, "naked Lua" functions are not permitted any more
- Bytecode is blocked in all versions of Lua except PUC Rio Lua 5.1
- The library throws an error when attempting to use quotas in LuaJIT
- Environments are now strictly read-only
- Environments can have metatables with indexes, and they are respected
- Environments can override the base environment

# v0.5.0 (2013)

Initial version


spec/sandbox_spec.lua
local sandbox = require 'sandbox'

describe('sandbox.run', function()

  describe('when handling base cases', function()

    it('can run harmless strings', function()
      local r = sandbox.run("return 'hello'")
      assert.equal(r, 'hello')
    end)

    if sandbox.bytecode_blocked then
      it('rejects bytecode', function()
        local fn = function() end
        assert.error(function() sandbox.run(string.dump(fn)) end)
      end)
    else
      it('accepts bytecode (PUC Rio 5.1)', function()
        local fn = function() end
        assert.has.no.error(function() sandbox.run(string.dump(fn)) end)
      end)
    end

    it('has access to safe methods', function()
      assert.equal(10,      sandbox.run("return tonumber('10')"))
      assert.equal('HELLO', sandbox.run("return string.upper('hello')"))
      assert.equal(1,       sandbox.run("local a = {3,2,1}; table.sort(a); return a[1]"))
      assert.equal(10,      sandbox.run("return math.max(1,10)"))
    end)

    it('does not allow access to not-safe stuff', function()
      assert.error(function() sandbox.run('return setmetatable({}, {})') end)
      assert.error(function() sandbox.run('return string.rep("hello", 5)') end)
    end)

    it('does return multiple values', function()
      local result = { sandbox.run("return 'hello', 'world'") }
      assert.same({ 'hello', 'world' }, result)
    end)
  end)

  describe('when handling string.rep', function()
    it('does not allow pesky string:rep', function()
      assert.error(function() sandbox.run('return ("hello"):rep(5)') end)
    end)

    it('restores the value of string.rep', function()
      sandbox.run("")
      assert.equal('hellohello', string.rep('hello', 2))
    end)

    it('restores string.rep even if there is an error', function()
      assert.error(function() sandbox.run("error('foo')") end)
      assert.equal('hellohello', string.rep('hello', 2))
    end)

    it('passes parameters to the code', function()
      assert.equal(sandbox.run("local a, b = ...; return a + b", {}, 1,2), 3)
    end)
  end)


  describe('when the sandboxed code tries to modify the base environment', function()

    it('does not allow modifying the modules', function()
      assert.error(function() sandbox.run("string.foo = 1") end)
      assert.error(function() sandbox.run("string.char = 1") end)
    end)

    it('does not persist modifications of base functions', function()
      sandbox.run('error = function() end')
      assert.error(function() sandbox.run("error('this should be raised')") end)
    end)

    it('does not persist modification to base functions even when they are provided by the base env', function()
      local env = {['next'] = 'hello'}
      sandbox.run('next = "bye"', { env=env })
      assert.equal(env['next'], 'hello')
    end)
  end)


  if sandbox.quota_supported then
    describe('when given infinite loops', function()
      it('throws an error with infinite loops', function()
        assert.error(function() sandbox.run("while true do end") end)
      end)

      it('restores string.rep even after a while true', function()
        assert.error(function() sandbox.run("while true do end") end)
        assert.equal('hellohello', string.rep('hello', 2))
      end)

      it('accepts a quota param', function()
        assert.has_no.errors(function() sandbox.run("for i=1,100 do end") end)
        assert.error(function() sandbox.run("for i=1,100 do end", {quota = 20}) end)
      end)

      it('does not use quotes if the quote param is false', function()
        assert.has_no.errors(function() sandbox.run("for i=1,1000000 do end", {quota = false}) end)
      end)
    end)
  else
    it('throws an error when trying to use the quota option in an unsupported environment (LuaJIT)', function()
      assert.error(function() sandbox.run("", {quota = 20}) end)
    end)
  end


  describe('when given an env option', function()
    it('is available on the sandboxed env as the _G variable', function()
      local env = {foo = 1}
      assert.equal(1, sandbox.run("return foo", {env = env}))
      assert.equal(1, sandbox.run("return _G.foo", {env = env}))
    end)

    it('does not hide base env', function()
      assert.equal('HELLO', sandbox.run("return string.upper(foo)", {env = {foo = 'hello'}}))
    end)

    it('cannot modify the env', function()
      local env = {foo = 1}
      sandbox.run("foo = 2", {env = env})
      assert.equal(env.foo, 1)
    end)

    it('uses the env metatable, if it exists', function()
      local env1 = { foo = 1 }
      local env2 = { bar = 2 }
      setmetatable(env2, { __index = env1 })
      assert.equal(3, sandbox.run("return foo + bar", { env = env2 }))
    end)

    it('can override the base env', function()
      local env = { tostring = function(x) return "hello " .. x end }
      assert.equal("hello peter", sandbox.run("return tostring('peter')", { env = env }))
    end)

    it('can override the base env with false', function()
      local env = { tostring = false }
      assert.equal(false, sandbox.run("return tostring", { env = env }))
    end)
  end)

end)


sandbox.lua
local sandbox = {
  _VERSION      = "sandbox 0.5",
  _DESCRIPTION  = "A pure-lua solution for running untrusted Lua code.",
  _URL          = "https://github.com/kikito/sandbox.lua",
  _LICENSE      = [[
    MIT LICENSE

    Copyright (c) 2021 Enrique García Cota

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  ]],

}

-- quotas don't work in LuaJIT since debug.sethook works differently there
local quota_supported = type(_G.jit) == "nil"
sandbox.quota_supported = quota_supported

-- PUC-Rio Lua 5.1 does not support deactivation of bytecode
local bytecode_blocked = _ENV or type(_G.jit) == "table"
sandbox.bytecode_blocked = bytecode_blocked

-- The base environment is merged with the given env option (or an empty table, if no env provided)
--
local BASE_ENV = {}

-- List of unsafe packages/functions:
--
-- * string.rep: can be used to allocate millions of bytes in 1 operation
-- * {set|get}metatable: can be used to modify the metatable of global objects (strings, integers)
-- * collectgarbage: can affect performance of other systems
-- * dofile: can access the server filesystem
-- * _G: It has access to everything. It can be mocked to other things though.
-- * load{file|string}: All unsafe because they can grant acces to global env
-- * raw{get|set|equal}: Potentially unsafe
-- * module|require|module: Can modify the host settings
-- * string.dump: Can display confidential server info (implementation of functions)
-- * math.randomseed: Can affect the host sytem
-- * io.*, os.*: Most stuff there is unsafe, see below for exceptions


-- Safe packages/functions below
([[

_VERSION assert error    ipairs   next pairs
pcall    select tonumber tostring type unpack xpcall

coroutine.create coroutine.resume coroutine.running coroutine.status
coroutine.wrap   coroutine.yield

math.abs   math.acos math.asin  math.atan math.atan2 math.ceil
math.cos   math.cosh math.deg   math.exp  math.fmod  math.floor
math.frexp math.huge math.ldexp math.log  math.log10 math.max
math.min   math.modf math.pi    math.pow  math.rad   math.random
math.sin   math.sinh math.sqrt  math.tan  math.tanh

os.clock os.difftime os.time

string.byte string.char  string.find  string.format string.gmatch
string.gsub string.len   string.lower string.match  string.reverse
string.sub  string.upper

table.insert table.maxn table.remove table.sort

]]):gsub('%S+', function(id)
  local module, method = id:match('([^%.]+)%.([^%.]+)')
  if module then
    BASE_ENV[module]         = BASE_ENV[module] or {}
    BASE_ENV[module][method] = _G[module][method]
  else
    BASE_ENV[id] = _G[id]
  end
end)

local function protect_module(module, module_name)
  return setmetatable({}, {
    __index = module,
    __newindex = function(_, attr_name, _)
      error('Can not modify ' .. module_name .. '.' .. attr_name .. '. Protected by the sandbox.')
    end
  })
end

('coroutine math os string table'):gsub('%S+', function(module_name)
  BASE_ENV[module_name] = protect_module(BASE_ENV[module_name], module_name)
end)

-- auxiliary functions/variables

local string_rep = string.rep

local function sethook(f, key, quota)
  if type(debug) ~= 'table' or type(debug.sethook) ~= 'function' then return end
  debug.sethook(f, key, quota)
end

local function cleanup()
  sethook()
  string.rep = string_rep -- luacheck: no global
end

-- Public interface: sandbox.protect
function sandbox.protect(code, options)
  options = options or {}

  local quota = false
  if options.quota and not quota_supported then
    error("options.quota is not supported on this environment (usually LuaJIT). Please unset options.quota")
  end
  if options.quota ~= false then
    quota = options.quota or 500000
  end

  assert(type(code) == 'string', "expected a string")

  local passed_env = options.env or {}
  local env = {}
  for k, v in pairs(BASE_ENV) do
    local pv = passed_env[k]
    if pv ~= nil then
      env[k] = pv
    else
      env[k] = v
    end
  end
  setmetatable(env, { __index = options.env })
  env._G = env

  local f
  if bytecode_blocked then
    f = assert(load(code, nil, 't', env))
  else
    f = assert(loadstring(code))
    setfenv(f, env)
  end

  return function(...)

    if quota and quota_supported then
      local timeout = function()
        cleanup()
        error('Quota exceeded: ' .. tostring(quota))
      end
      sethook(timeout, "", quota)
    end

    string.rep = nil -- luacheck: no global

    local t = table.pack(pcall(f, ...))

    cleanup()

    if not t[1] then error(t[2]) end

    return table.unpack(t, 2, t.n)
  end
end

-- Public interface: sandbox.run
function sandbox.run(code, options, ...)
  return sandbox.protect(code, options)(...)
end

-- make sandbox(f) == sandbox.protect(f)
setmetatable(sandbox, {__call = function(_,code,o) return sandbox.protect(code,o) end})

return sandbox
